### PR TITLE
Allow dial retries for the downstream proxy connection

### PR DESCRIPTION
**Pull request by ctophs**

### CODE REVIEW

**Comment by L11R** [comment]
@ctophs looks ok for me. Does it resolve your problem?

File: internal/proxy/proxy.go [hunk 1]
```diff
@@ -259,11 +259,20 @@ func (p *Proxy) https(rw http.ResponseWriter, req *http.Request) {
 func (p *Proxy) connectAndCopy(conn net.Conn, brw *bufio.ReadWriter, req *http.Request, reconnected bool) {
 	logger := req.Context().Value(LogEntryCtx).(*zap.Logger)
 
+	retries := 0
+proxy_dial_retry:
 	// Open connection with downstream proxy
 	pconn, err := net.DialTimeout("tcp", p.config.DownstreamProxyURL.Host, p.config.Timeouts.DownstreamProxy.DialTimeout)
 	if err != nil {
+		if retries < p.config.DownstreamProxyDialRetries {
+			logger.Error("connection to downstream proxy failed.", zap.Error(err))
```
Comment (L11R) [inline internal/proxy/proxy.go:268]: If i remembered it correctly, log messages use Upper case. Can you edit it if true? For the log consistency.

File: internal/proxy/proxy.go [hunk 2]
```diff
@@ -259,11 +259,20 @@ func (p *Proxy) https(rw http.ResponseWriter, req *http.Request) {
 func (p *Proxy) connectAndCopy(conn net.Conn, brw *bufio.ReadWriter, req *http.Request, reconnected bool) {
 	logger := req.Context().Value(LogEntryCtx).(*zap.Logger)
 
+	retries := 0
+proxy_dial_retry:
```
Comment (L11R) [inline internal/proxy/proxy.go:263]: Go name conventions suggest to use CamelCase for labels too.
https://go.dev/doc/effective_go#mixed-caps

**Comment by ctophs** [comment]
yes, it does. It's just a grace time of a few seconds when wsl-vpnkit messes with the windows networking so with a retry of 3 I'm fine. Defauts to 0 so it shouldn't affect anybody without that odd problem

File: internal/proxy/proxy.go [hunk 3]
```diff
@@ -259,11 +259,20 @@ func (p *Proxy) https(rw http.ResponseWriter, req *http.Request) {
...
 	// Open connection with downstream proxy
 	pconn, err := net.DialTimeout("tcp", p.config.DownstreamProxyURL.Host, p.config.Timeouts.DownstreamProxy.DialTimeout)
 	if err != nil {
+		if retries < p.config.DownstreamProxyDialRetries {
+			logger.Error("connection to downstream proxy failed.", zap.Error(err))
+			retries++
+			logger.Debug("downstream proxy dial retry", zap.Int("retries", retries))
```
Comment (L11R) [inline internal/proxy/proxy.go:270]: Same here!

**Review by L11R — CHANGES_REQUESTED** [review]

File: internal/proxy/proxy.go (hunk 1 above)
Comment (ctophs) [reply internal/proxy/proxy.go:268]: you mean first letter of the message should be uppercase?

internal/proxy/util.go:                                 c.logger.Error("cannot close write", zap.Error(err))
internal/proxy/proxy.go:                                logger.Error("Panic recovered", zap.Error(err))

File: internal/proxy/proxy.go (hunk 2 above)
Comment (ctophs) [reply internal/proxy/proxy.go]: changed to ProxyDialRetries

**Comment by L11R** [comment]
@ctophs oh, i guess line 275 became unreachable.

Let's make this code block like this:
```go
ProxyDialRetry:
	// Open connection with downstream proxy
	pconn, err := net.DialTimeout("tcp", p.config.DownstreamProxyURL.Host, p.config.Timeouts.DownstreamProxy.DialTimeout)
	if err != nil {
		if retries < p.config.DownstreamProxyDialRetries {
			logger.Error("Connection to downstream proxy failed.", zap.Error(err))
			retries++
			logger.Debug("Downstream proxy dial retry", zap.Int("retries", retries))
			goto ProxyDialRetry
		}

		httpsErrorHijackedHandler(brw, req, fmt.Errorf("cannot connect to downstream proxy: %w", err))
	}
```

**Comment by ctophs** [comment]
maybe we should insert a time.Sleep(1 * time.Second) inbetween the retries

**Comment by L11R** [comment]
@ctophs i've released pre-release. Please test it! I've also updated all the deps.
https://github.com/L11R/escobar/releases/tag/v3.0.1

**Comment by ctophs** [comment]
works for me. Thank you very much!

**Pull request merged**